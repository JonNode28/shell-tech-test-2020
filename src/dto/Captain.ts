export default class Captain {
  name: string;
  constructor(name: string) {
    this.name = name;
  }
}