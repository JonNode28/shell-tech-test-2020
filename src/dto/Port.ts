export default class Port {
  name: string;
  constructor(name: string) {
    this.name = name;
  }

}