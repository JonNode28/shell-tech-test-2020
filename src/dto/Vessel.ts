export default class Vessel {
  name: string;
  constructor(name: string) {
    this.name = name;
  }
}