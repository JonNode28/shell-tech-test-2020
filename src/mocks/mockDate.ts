const mockDate = (expected: Date) => {
  const _Date = Date;

  function MockDate(mockOverride?: Date | number) {
    return new _Date(mockOverride || expected)
  }

  MockDate.UTC = _Date.UTC;
  MockDate.parse = _Date.parse;
  MockDate.now = () => expected.getTime();
  MockDate.prototype = _Date.prototype;

  global.Date = MockDate as any;

  return _Date;
};

export default mockDate;

export const resetDate = (originalDate: any) => {
  global.Date = Date;
};