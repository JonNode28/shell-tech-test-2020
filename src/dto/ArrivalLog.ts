import Captain from './Captain';
import Vessel from "./Vessel";
import Port from "./Port";

export default class ArrivalLog {
  readonly captain: Captain;
  readonly vessel: Vessel;
  readonly port: Port;
  readonly timestamp: Date;
  constructor(captain: Captain, vessel: Vessel, port: Port) {
    this.captain = captain;
    this.vessel = vessel;
    this.port = port;
  }
}