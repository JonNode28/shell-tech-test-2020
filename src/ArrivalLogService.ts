import ArrivalLog from "./dto/ArrivalLog";

export default class ArrivalLogService {
  arrivalLogs: Array<ArrivalLog> = new Array<ArrivalLog>();
  timestamp: Date = new Date();
  save(arrivalLog: ArrivalLog) {
    this.arrivalLogs.push(arrivalLog);
  }
  list(captainName: string): Array<ArrivalLog>{
    const result = captainName ? this.arrivalLogs
      .filter(log => log.captain.name === captainName) : this.arrivalLogs;

    return result
      .sort((a, b) => b.timestamp.getTime() - a.timestamp.getTime());
  }
}