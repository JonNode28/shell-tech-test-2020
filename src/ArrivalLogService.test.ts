import ArrivalLogService from "./ArrivalLogService";
import ArrivalLog from "./dto/ArrivalLog";
import Captain from "./dto/Captain";
import Vessel from "./dto/Vessel";
import Port from "./dto/Port";
import mockDate, { resetDate } from './mocks/mockDate';

describe('ArrivalLogService.test.ts', () => {
  let sut;
  beforeEach(() => {
    sut = new ArrivalLogService();
  });

  it(`should save an arrival log so that it's retrievable`, () => {
    const originalDate = mockDate(new Date(0));
    sut.save(new ArrivalLog(new Captain('Janeway'), new Vessel('Voyager'), new Port('Earth')));
    const result = sut.list('Janeway');
    expect(result).toMatchSnapshot();
    resetDate(originalDate);
  });

  it('should list all logs when no captain name is supplied', () => {
    const originalDate = mockDate(new Date(1));
    sut.save(new ArrivalLog(new Captain('Janeway'), new Vessel('Voyager'), new Port('Earth')));
    sut.save(new ArrivalLog(new Captain('Janeway'), new Vessel('Voyager'), new Port('Mars')));
    sut.save(new ArrivalLog(new Captain('Janeway'), new Vessel('Voyager'), new Port('Jupiter')));
    const result = sut.list();
    expect(result).not.toBeNull();
    expect(result).toHaveLength(3);
    expect(result).toMatchSnapshot();
    resetDate(originalDate);
  });

  it('should list all logs ordered descending by timestamp', () => {
    const originalDate = mockDate(new Date(0));
    sut.save(new ArrivalLog(new Captain('Janeway'), new Vessel('Voyager'), new Port('Earth')));
    mockDate(new Date(2));
    sut.save(new ArrivalLog(new Captain('Janeway'), new Vessel('Voyager'), new Port('Mars')));
    mockDate(new Date(1));
    sut.save(new ArrivalLog(new Captain('Janeway'), new Vessel('Voyager'), new Port('Jupiter')));
    const result = sut.list();
    expect(result).not.toBeNull();
    expect(result).toHaveLength(3);
    expect(result[0].port.name).toEqual('Earth');
    expect(result[1].port.name).toEqual('Jupiter');
    expect(result[2].port.name).toEqual('Mars');
    resetDate(originalDate);
  });

});