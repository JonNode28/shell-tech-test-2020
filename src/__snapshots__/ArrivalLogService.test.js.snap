// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArrivalLogService.test.ts should list all logs when no captain name is supplied 1`] = `
Array [
  ArrivalLog {
    "captain": Captain {
      "name": "Janeway",
    },
    "port": Port {
      "name": "Earth",
    },
    "timestamp": 1970-01-01T00:00:00.001Z,
    "vessel": Vessel {
      "name": "Voyager",
    },
  },
  ArrivalLog {
    "captain": Captain {
      "name": "Janeway",
    },
    "port": Port {
      "name": "Mars",
    },
    "timestamp": 1970-01-01T00:00:00.001Z,
    "vessel": Vessel {
      "name": "Voyager",
    },
  },
  ArrivalLog {
    "captain": Captain {
      "name": "Janeway",
    },
    "port": Port {
      "name": "Jupiter",
    },
    "timestamp": 1970-01-01T00:00:00.001Z,
    "vessel": Vessel {
      "name": "Voyager",
    },
  },
]
`;

exports[`ArrivalLogService.test.ts should save an arrival log so that it's retrievable 1`] = `
Array [
  ArrivalLog {
    "captain": Captain {
      "name": "Janeway",
    },
    "port": Port {
      "name": "Earth",
    },
    "timestamp": 1970-01-01T00:00:00.000Z,
    "vessel": Vessel {
      "name": "Voyager",
    },
  },
]
`;
